src/rt/cpp/cown.h

```diff
@@ -0,0 +1,165 @@
+// Copyright Microsoft and Project Verona Contributors.
+// SPDX-License-Identifier: MIT
+
+#include <functional>
+#include <tuple>
+#include <utility>
+#include <verona.h>
+
+class cown_ptr_trait
+{
+private:
+  cown_ptr_trait() {}
+
+  template<typename T>
+  friend class cown_ptr;
+};
+
+/**
+ * Smart pointer to represent shared access to a cown.
+ * Can only be used asychronously with `when` to get
+ * underlying access.
+ */
+template<typename T>
+class cown_ptr : cown_ptr_trait
+{
+private:
+  /**
+   * Internal Verona runtime cown for this type.
+   */
+  class ActualCown : public VCown<ActualCown>
+  {
+  private:
+    T value;
+
+  public:
+    ActualCown(T&& t) : value(std::forward<T>(t)) {}
+
+    template<typename TT>
+    friend class acquired_cown;
+  };
+
+  /**
+   * Internal Verona runtime cown for this type.
+   */
+  ActualCown* allocated_cown = nullptr;
+
+  /**
+   * Accesses the internal Verona runtime cown for this handle.
+   */
+  Cown* underlying_cown()
+  {
+    return allocated_cown;
+  }
+
+  /**
+   * Construct a new cown ptr object, actually allocates a runtime cown.
+   *
+   * This is internal, and the `make_cown` below is the public interface,
+   * which has better behaviour for implicit template arguments.
+   */
+  cown_ptr(T&& t) : allocated_cown(new ActualCown(std::forward<T>(t))) {}
+
+public:
+  /**
+   * Copy an existing cown ptr.  Shares the underlying cown.
+   */
+  cown_ptr(const cown_ptr& other)
+  {
+    allocated_cown = other.allocated_cown;
+    verona::rt::Cown::acquire(allocated_cown);
+  }
+
+  /**
+   * Move an existing cown ptr.  Does not create a new cown,
+   * and is more efficient than copying, as it does not need
+   * to perform reference count operations.
+   */
+  cown_ptr(cown_ptr&& other)
+  {
+    allocated_cown = other.allocated_cown;
+    other.allocated_cown = nullptr;
+  }
+
+  ~cown_ptr()
+  {
+    // Condition to handle moved cown ptrs.
+    // And a defence against the destructor being called twice.
+    if (allocated_cown != nullptr)
+    {
+      auto& alloc = verona::rt::ThreadAlloc::get();
+      verona::rt::Cown::release(alloc, allocated_cown);
+      allocated_cown = nullptr;

```

Comment:
Note that this will almost certainly be optimised away above O0 because it is a dead store according to the abstract machine (it is UB to read from a field after the destructor has finished).